### PR TITLE
feat: promote product tile with ribbons and banners

### PR DESCRIPTION
**Pull request by InnaBoitsun**

### CODE REVIEW

File: src/Orckestra.Composer.CompositeC1/Context/ProductTileConfigurationContext.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            DataQueryService = dataQueryService;
+            PromotionalRibbonCache = cacheService.GetStoreWithDependencies<string, List<IPromotionalRibbonConfiguration>>("Promotional Ribbons",
+                new CacheDependentEntry<IPromotionalRibbonConfiguration>());
+
+            PromotionalBannerCache = cacheService.GetStoreWithDependencies<string, List<IPromotionalBannerConfiguration>>("Promotional Banners",
+                new CacheDependentEntry<IPromotionalBannerConfiguration>());
+        }
+        public List<IPromotionalRibbonConfiguration> GetPromotionalRibbonConfigurations()
+        {
+            if (_promotionalRibbonConfigurations == null)
```
Comment (neexite) [inline src/Orckestra.Composer.CompositeC1/Context/ProductTileConfigurationContext.cs:35]: Why do we not return PromotionalRibbonCache.GetOrAdd("PromotionalRibbonSettings", _ => LoadPromotionalRibbonSettings())
The class registered as Transistion and _promotionalRibbonConfigurations will be null everytime

File: src/Orckestra.Composer.CompositeC1/Context/ProductTileConfigurationContext.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            }
+
+            return promotionalRibbons;
+        }
+
+        public List<IPromotionalBannerConfiguration> GetPromotionalBannerConfigurations()
+        {
+            if (_promotionalBannerConfigurations == null)
+            {
+                _promotionalBannerConfigurations = PromotionalBannerCache.GetOrAdd("PromotionalBannerSettings", _ => LoadPromotionalBannerSettings());
```
Comment (neexite) [inline src/Orckestra.Composer.CompositeC1/Context/ProductTileConfigurationContext.cs:58]: gotta, why do we use _promotionalBannerConfigurations ?

File: src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            IProductTileConfigurationContext productTileConfigurationContext)
+        {
+            ProductTileConfigurationContext = productTileConfigurationContext ?? throw new ArgumentNullException(nameof(productTileConfigurationContext));
+            LocalizationProvider = localizationProvider ?? throw new ArgumentNullException(nameof(localizationProvider));
+        }
+
+        public virtual Dictionary<string, string> BuildProductBadgeValues(string keys, string lookupDisplayNames)
+        {
+            var productBadgesNames = lookupDisplayNames != null ? lookupDisplayNames.Split(',').ToList() : new List<string>();
+            var productBadgeKeys = keys != null ? keys.Split('|').ToList() : new List<string>();
```
Comment (neexite) [inline src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs:27]: A little bit shorter construction

```
            var productBadgesNames = lookupDisplayNames?.Split(',').ToList() ?? new List<string>();
            var productBadgeKeys = keys?.Split('|').ToList() ?? new List<string>();
```

File: src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+            if (!productBadgeKeys.Any())
```
Comment (neexite) [inline src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs:29]: we can put upper if(keys == null) return null, and didn't check for null for productBadgeKeys

File: src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                return null;
+            }
+
+            var productBadgeValues = new Dictionary<string, string>();
+
+            for (var i = 0; i < productBadgeKeys.Count; i++)
+            {
+                if (!productBadgeValues.ContainsValue(productBadgeKeys[i]))
+                    productBadgeValues.Add(productBadgeKeys[i], productBadgesNames[i]);
```
Comment (neexite) [inline src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs:39]: we don't check that productBadgesNames have the same count as productBadgeKeys, so IndexOutOfRangeException  can happen

File: src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            var productBadgeKeys = keys != null ? keys.Split('|').ToList() : new List<string>();
+
+            if (!productBadgeKeys.Any())
+            {
+                return null;
+            }
+
+            var productBadgeValues = new Dictionary<string, string>();
+
+            for (var i = 0; i < productBadgeKeys.Count; i++)
```
Comment (neexite) [inline src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs:36]: You can use Zip to iterate two lists
          ```
  foreach (var tuple in productBadgeKeys.Zip(productBadgesNames, (x, y) => (Key:x, Value: y)))
            {
                productBadgeValues[tuple.Key] = tuple.Value;
            }
```

File: src/Orckestra.Composer.Product/ViewModels/RelatedProductViewModel.cs [hunk 7]
```diff
@@ -29,5 +30,61 @@ public sealed class RelatedProductViewModel : BaseProductViewModel
...
+        /// </summary>
+        [Lookup(LookupType.Product, "PromotionalRibbon")]
+        public string PromotionalRibbon { get; set; }
+
+        /// <summary>
+        /// Gets or sets the product Promotional Ribbon Background Color
+        /// </summary>
+        public string PromotionalRibbonBackgroundColor { get; set; }
+
+        // <summary>
```
Comment (neexite) [inline src/Orckestra.Composer.Product/ViewModels/RelatedProductViewModel.cs:61]: missing 1 dash

File: src/Orckestra.Composer.Product/ViewModels/RelatedProductViewModel.cs [hunk 8]
```diff
@@ -29,5 +30,61 @@ public sealed class RelatedProductViewModel : BaseProductViewModel
...
+        public string ProductBadgesLookup { get; set; }
+
+        public Dictionary<string, string> ProductBadgeValues { get; set; }
+
+        /// <summary>
+        /// Gets or sets the product Promotional Ribbon Value
+        /// </summary>
+        [MapTo("PromotionalRibbon")]
+        public string PromotionalRibbonValue { get; set; }
+        /// <summary>
```
Comment (neexite) [inline src/Orckestra.Composer.Product/ViewModels/RelatedProductViewModel.cs:50]: missing new line

File: src/Orckestra.Composer.Product/ViewModels/RelatedProductViewModel.cs [hunk 9]
```diff
@@ -29,5 +30,61 @@ public sealed class RelatedProductViewModel : BaseProductViewModel
...
+        /// </summary>
+        [Lookup(LookupType.Product, "PromotionalBanner")]
+        public string PromotionalBanner { get; set; }
+
+        /// <summary>
+        /// Gets or sets the product Promotional Banner Background Color
+        /// </summary>
+        public string PromotionalBannerBackgroundColor { get; set; }
+
+        // <summary>
```
Comment (neexite) [inline src/Orckestra.Composer.Product/ViewModels/RelatedProductViewModel.cs:84]: missing 1 dash

File: src/Orckestra.Composer.Search/ViewModels/ProductSearchViewModel.cs [hunk 10]
```diff
@@ -20,5 +21,42 @@ public class ProductSearchViewModel : BaseProductViewModel
...
+        /// Gets or sets the product Promotional Ribbon
+        /// </summary>
+        public string PromotionalRibbon { get; set; }
+
+        /// <summary>
+        /// Gets or sets the product Promotional Ribbon Background Color
+        /// </summary>
+        public string PromotionalRibbonBackgroundColor { get; set; }
+
+        // <summary>
```
Comment (neexite) [inline src/Orckestra.Composer.Search/ViewModels/ProductSearchViewModel.cs:41]: missing 1 dash

File: src/Orckestra.Composer.Search/ViewModels/ProductSearchViewModel.cs [hunk 11]
```diff
@@ -20,5 +21,42 @@ public class ProductSearchViewModel : BaseProductViewModel
...
+        /// Gets or sets the product Promotional Banner
+        /// </summary>
+        public string PromotionalBanner { get; set; }
+
+        /// <summary>
+        /// Gets or sets the product Promotional Banner Background Color
+        /// </summary>
+        public string PromotionalBannerBackgroundColor { get; set; }
+
+        // <summary>
```
Comment (neexite) [inline src/Orckestra.Composer.Search/ViewModels/ProductSearchViewModel.cs:56]: missing 1 dash

**Review by neexite — CHANGES_REQUESTED** [review]

File: src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs (hunk 5 above)
Comment (InnaBoitsun) [reply src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs]: I added unit tests to validate all cases and did changes

**Review by neexite — APPROVED** [review]

**Pull request merged**